// common/Gruntfile.js
module.exports = function(grunt) {

    var tsFiles = ['src/**/*.ts'];
    var jsFiles = ['dist/**/*.js'];
    var configFiles= ['Gruntfile.js','tslint.json','tsconfig.json','typings.json','package.json','nodemon.json'];

    grunt.initConfig({
        concurrent: {
            start:['watch','nodemon:dev'],
            ts:['tslint','ts'],
            tsNoLint:['ts'],
            startLiveTest:['watch'],
            options: {
                logConcurrentOutput: true
            }
        },
        nodemon:{
            dev: {
                script: 'dist/main/index.js',
                options:{
                    watch: [jsFiles],
                    interrupt: true
                }
            }
        },
        mochaTest: {
            test: {
                options: {
                    reporter: 'spec',
                    quiet: false, // Optionally suppress output to standard out (defaults to false)
                    clearRequireCache: false // Optionally clear the require cache before running tests (defaults to false)
                },
                src: ['dist/test/unit/**/*.js']
            }
        },
        tslint: {
            options: {
                // can be a configuration object or a filepath to tslint.json
                configuration: "tslint.json"
            },
            files: {
                src: tsFiles
            }
        },
        ts: {
            dev: {
                tsconfig: true,
                src: tsFiles,
                outDir: 'dist'
            }
        },
        mocha_istanbul: {
            coveralls: {
                src: ['dist/test/unit'], // multiple folders also works
                options: {
                    coverage:false, // this will make the grunt.event.on('coverage') event listener to be triggered
                    coverageFolder: 'dist/test/coverage',
                    check: {

                    },
                    root: './dist/main', // define where the cover task should consider the root of libraries that are covered by tests
                    reportFormats: ['json','lcovonly'],
                    print: 'summary'

                }
            }
        },
        remapIstanbul: {
            build: {
                src: 'dist/test/coverage/coverage-final.json',
                options: {
                    reports: {
                        'html': 'dist/test/coverage/html-report',
                        'json': 'dist/test/coverage/coverage-final.json'
                    }
                }
            }
        },
        watch: {
            ts: {
                files :[tsFiles, configFiles],
                tasks:['concurrent:ts'],
                options: {
                    interrupt: true
                }
            },
            js : {
                files :[jsFiles],
                tasks:['mochaTest'],
                options: {
                    interrupt: true
                }
            }
        }
    });


    //Load Grunt tasks.
    grunt.loadNpmTasks('remap-istanbul');
    require('load-grunt-tasks')(grunt);

    /*Available grunt tasks*/
    grunt.registerTask('default', ['concurrent:start']);

    grunt.registerTask('live-test', ['concurrent:startLiveTest']);

    grunt.registerTask('test', ['concurrent:ts','mocha_istanbul','remapIstanbul:build']);
    grunt.registerTask('testNoLint', ['concurrent:tsNoLint', 'mocha_istanbul', 'remapIstanbul:build'])

    grunt.registerTask('build', ['concurrent:ts']);
};
